src/com/twitter/intellij/pants/components/impl/FastpassUpdater.java

```diff
@@ -0,0 +1,254 @@
+          .forEach(project -> extractFastpassData(project)
+            .ifPresent(data -> {
+              if (!data.version.equals(systemVersion)) {
+                showUpdateNotification(project, systemVersion, data);
+              }
+            })));
+    }
+  }
+
+  private static void showUpdateNotification(Project project, String systemVersion, FastpassData data) {
+    if (!hasNotification(project)) {
+      Notification notification = new NotificationGroup(PantsConstants.PANTS, NotificationDisplayType.STICKY_BALLOON, true)
+        .createNotification(
+          NOTIFICATION_TITLE,
+          "<a href='" + NOTIFICATION_HREF + "'>Update</a> fastpass to version: " + systemVersion,
+          NotificationType.INFORMATION,
+          new NotificationListener.Adapter() {
+            @Override
+            protected void hyperlinkActivated(@NotNull Notification notification, @NotNull HyperlinkEvent e) {
+              if (NOTIFICATION_HREF.equals(e.getDescription())) {
+                updateFastpassVersion(project, data);
+              }
+              notification.expire();
+            }
+          }
+        );
+
+      notification.notify(project);
+    }
+  }
+
+  private static boolean hasNotification(Project project) {
+    ArrayList<Notification> notifications = EventLog.getLogModel(project).getNotifications();
+    return notifications.stream().anyMatch(s -> s.getTitle().equals(NOTIFICATION_TITLE));
+  }
+
+  private static void updateFastpassVersion(Project project, FastpassData data) {
+    ProgressManager.getInstance().run(new Task.Backgroundable(project, "Updating fastpass version") {
+      @Override
+      public void run(@NotNull ProgressIndicator indicator) {
+        indicator.setText("Exiting bloop");
+        exitBloop();
+        indicator.setText("Running fastpass refresh");
+        boolean ok = runFastpassRefresh(data, project);
+        if (ok) {
+          indicator.setText("Refreshing BSP project after fastpass update");
+          PantsUtil.refreshAllProjects(project);
+        }
+      }
+    });
+  }
+
+  private static void exitBloop() {
+    try {
+      new ProcessBuilder(BLOOP_PATH, "exit").start().waitFor();
+    }
+    catch (Exception e) {
+      LOG.warn(e);
+    }
+  }
+
+  private static boolean runFastpassRefresh(FastpassData data, Project project) {
+    try {
+      GeneralCommandLine commandLine = PantsUtil.defaultCommandLine(project);
+      commandLine.setExePath(FASTPASS_PATH);
+      commandLine.addParameters(
+        "refresh",
+        "--intellij",
+        "--intellijLauncher", "echo", // to avoid opening project again
+        data.projectName
+      );
+      Process refresh = commandLine.createProcess();
+      refresh.waitFor();
+      return refresh.exitValue() == 0;
+    }
+    catch (Exception e) {
+      LOG.warn(e);
+      return false;
+    }
+  }
+
+  private static Optional<FastpassData> extractFastpassData(Project project) {
```

Comment:
I just noticed that this method relies on the implementation details of `bloop.settings.json` that may change in the future. Can we first add a check that tries to read the `fastpassVersion` field in `.bsp/bloop.json`? This field is currently not generated by fastpass so it won't be used now, but we can add it separately to Metals which would make this logic less fragile.